### PR TITLE
Build and test mergely on JDK 16 (on Travis)

### PR DESCRIPTION
**Pull request by lrytz**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -8,120 +8,147 @@ stages:
...
-          - bundle exec jekyll build -s spec/ -d build/spec
-          - export JEKYLL_ENV=spec-pdf
-          - bundle exec jekyll build -s spec/ -d build/spec-pdf
-          - ./scripts/generate-spec-pdf.sh
-        after_success:
-          - ./scripts/travis-publish-spec.sh
+  include:
+    - stage: build
+      if: type != pull_request AND repo = scala/scala
+      name: publish to scala-integration or sonatype
```
Comment (dwijnand) [inline .travis.yml]: What does it use to decide to publish to sonatype?  And I think the bootstrap part is useful to know, reasons for removing?

File: .travis.yml [hunk 2]
```diff
@@ -8,120 +8,147 @@ stages:
   - build
   - test
 
+templates:
```
Comment (dwijnand) [inline .travis.yml]: Assuming I've understood correctly, could you leave a little comment to say that this tag is our invention to reuse these config in jobs.include.

File: .travis.yml [hunk 3]
```diff
@@ -8,120 +8,147 @@ stages:
...
+  pr-jdk8: &pr-jdk8
+    if: type = pull_request OR repo != scala/scala
+
+  push-jdk16: &push-jdk16
+    if: type = push AND repo = scala/scala
```
Comment (dwijnand) [inline .travis.yml]: So build on every merge (and tags - and sneaky commit pushes).  Which means 3 extra jobs every merge.  I wouldn't mind avoiding that overhead on PR merges and having this be a fixed daily +3 jobs, by having these build nightly instead.

**Review by dwijnand — CHANGES_REQUESTED** [review]
Mostly LGTM.  Just a tiny request, question, and a "let's not do it this way" pushback. 😄

**Comment by som-snytt** [comment]
I have a couple of test failures on jdk16... I mentioned it on 3613.
```
test/files/neg/macro-invalidret-neg.log
test/files/run/t3613-run.log
```

**Comment by SethTisue** [comment]
is the idea here that if we test on 16, we don't also need to test on 11?

File: .travis.yml (hunk 1 above)
Comment (lrytz) [reply .travis.yml]: if `SCALA_VER_BASE` is defined publishing goes to sonatype
https://github.com/scala/scala/blob/52cdad95982e0bf05d00a0caaaba7d6d2b490a2d/scripts/bootstrap_fun#L35-L48

i'll put back the `bootstrap`

File: .travis.yml (hunk 3 above)
Comment (lrytz) [reply .travis.yml]: That's fine with me 👍

**Comment by lrytz** [comment]
> is the idea here that if we test on 16, we don't also need to test on 11?

I'd say that's fine, since we have a community build on 11?

**Comment by SethTisue** [comment]
> I'd say that's fine, since we have a community build on 11?

I don't have a strong feeling either way.

(In the community build, I chose to have both 11 and 16 builds, but I chose that mainly because several important upstream maintainers don't support 15 or 16, so the 11 build tests noticeably more stuff. Here in this repo, that doesn't really apply, so 16 only is probably fine. On the other hand, doing 11 too wouldn't be especially costly. 🤷 )

**Comment by som-snytt** [comment]
On my lunch break, I am PRing removal of use of ThreadGroup in testkit, which is day job adjacent.

My other innovative contribution is to pronounce PR as purr.

**Comment by lrytz** [comment]
The JDK 16 build went green too: https://travis-ci.com/github/scala/scala/builds/223846952

(I pushed to a scala/scala branch for testing: https://github.com/scala/scala/tree/travis-jdk16)

re-review by @dwijnand

**Comment by lrytz** [comment]
(the github page summarizing the travis build incorrectly says `ADOPTOPENJDK=8`, https://github.com/scala/scala/runs/2408503535. if you look at the build log you see that the env is `ADOPTOPENJDK=16`, and `java -version` is correct)

**Review by dwijnand — APPROVED** [review]

**Comment by SethTisue** [comment]
@som-snytt https://twitter.com/SethTisue/status/1386797658988187648

**Comment by som-snytt** [comment]
@SethTisue whew, when I saw the notification while waiting at child's piano lesson, I wondered if something broke under `-Xsource:3` in the CB.

Thanks for reminding me that I still have to add the ThreadGroup commit to my PR for REPL pretty printing.

**Pull request merged**